plugins/templating/src/main/kotlin/templates/TemplateProcessor.kt

```diff
@@ -1,56 +1,92 @@
+import org.jetbrains.dokka.model.withDescendants
+import org.jetbrains.dokka.pages.RootPageNode
 import org.jetbrains.dokka.plugability.DokkaContext
 import org.jetbrains.dokka.plugability.plugin
 import org.jetbrains.dokka.plugability.query
+import org.jetbrains.dokka.plugability.querySingle
 import org.jsoup.nodes.Element
 import java.io.File
 
-interface TemplateProcessor {
-    fun process()
+interface TemplateProcessor
+
+interface SubmoduleTemplateProcessor : TemplateProcessor {
+    fun process(modules: List<DokkaConfiguration.DokkaModuleDescription>): TemplatingResult
+}
+
+interface MultiModuleTemplateProcessor : TemplateProcessor {
+    fun process(generatedPagesTree: RootPageNode)
 }
 
 interface TemplateProcessingStrategy {
     fun process(input: File, output: File): Boolean
     fun finish(output: File) {}
 }
 
-class DefaultTemplateProcessor(
+class DefaultSubmoduleTemplateProcessor(
     private val context: DokkaContext,
-): TemplateProcessor {
+) : SubmoduleTemplateProcessor {
 
-    private val strategies: List<TemplateProcessingStrategy> = context.plugin<TemplatingPlugin>().query { templateProcessingStrategy }
+    private val strategies: List<TemplateProcessingStrategy> =
+        context.plugin<TemplatingPlugin>().query { templateProcessingStrategy }
 
-    override fun process() = runBlocking(Dispatchers.Default) {
-        coroutineScope {
-            context.configuration.modules.forEach {
-                launch {
-                    it.sourceOutputDirectory.visit(context.configuration.outputDir.resolve(it.relativePathToOutputDirectory))
+    override fun process(modules: List<DokkaConfiguration.DokkaModuleDescription>) =
+        runBlocking(Dispatchers.Default) {
+            coroutineScope {
+                context.configuration.modules.fold(TemplatingResult()) { acc, module ->
+                    acc + module.sourceOutputDirectory.visit(context.configuration.outputDir.resolve(module.relativePathToOutputDirectory))
                 }
             }
         }
-        strategies.map { it.finish(context.configuration.outputDir) }
-        Unit
-    }
 
-    private suspend fun File.visit(target: File): Unit = coroutineScope {
-        val source = this@visit
-        if (source.isDirectory) {
-           target.mkdir()
-           source.list()?.forEach {
-               launch { source.resolve(it).visit(target.resolve(it)) }
-           }
-        } else {
-            strategies.first { it.process(source, target) }
+    private suspend fun File.visit(target: File, acc: TemplatingResult = TemplatingResult()): TemplatingResult =
+        coroutineScope {
+            val source = this@visit
+            if (source.isDirectory) {
+                target.mkdir()
+                val files = source.list().orEmpty()
+                val accWithSelf =
+                    if (files.contains("package-list") && files.size != 1) acc.copy(modules = acc.modules + source.name)
```

Comment:
I'd switch to path comparison for detecting top-level module and not generate empty modules at all